### ISSUE TITLE
[aot_eager] Accuracy error - BigBird

### ISSUE THREAD

**Issue by anijain2305** [pytorch/torchdynamo#1291]
Repro - `python benchmarks/huggingface.py --accuracy --float32 --backend=aot_eager --training --only=BigBird`

My investigation
* I ran accuracy minifier. Accuracy minifier takes the extracted Fx subgraph, runs that eagerly, and then with compiler_fn (aot_eager here) and compares the outputs. In this case, accuracy minifier did not find any offending subgraph.
* I recheck `eager` backend -- `torchdynamo.optimize("eager")`. It passed. So, something is wrong with `aot_eager` backend.

* At this point, I am thinking the problem must be in interstitial code between Fx subgraphs in Python bytecode. I thought that BigBird has numpy random calls, so maybe we are changing numpy rng state. But that was a dead end too as we reset numpy rng state before each run. `eager` backend is passing too, so interstitial code is ok.

* So, I just started arbitrarily falling back to eager here by adjusting the number of ops in Fx graph

https://github.com/pytorch/torchdynamo/blob/d6e2101148eb561aa8658765e0de1801e89f22eb/torchdynamo/optimizations/training.py#L32-L33

* After lots of hit and trial, I found that the this diff worked
~~~
diff --git a/torchdynamo/optimizations/training.py b/torchdynamo/optimizations/training.py
index 7f64c86c..b580180e 100644
--- a/torchdynamo/optimizations/training.py
+++ b/torchdynamo/optimizations/training.py
@@ -29,7 +29,8 @@ class AotAutogradStrategy(object):

     @classmethod
     def compile_fn(cls, gm: torch.fx.GraphModule, example_inputs):
-        if count_calls(gm.graph) < 2:
+        if count_calls(gm.graph) <= 2:
+            log.error(f"{gm}")
             return gm.forward  # no point for tiny graphs
         return cls(gm, example_inputs).verified_candidate()

~~~
And the additional skipped module is

~~~


def forward(self, _stack0_0_ : torch.Tensor):
    contiguous = _stack0_0_.contiguous();  _stack0_0_ = None
    view = contiguous.view(1, 1024, -1);  contiguous = None
    return (view,)

~~~

So, I have two questions
* What is the issue? I don't understand why skipping the above subgraph makes it work.
* Why accuracy minifier failed here?

**Comment by anijain2305** [comment]
cc @ezyang @eellison Any suggestions here?

**Comment by ezyang** [comment]
This looks like the same flavor as https://github.com/pytorch/torchdynamo/issues/1274

Why don't you try using https://github.com/pytorch/pytorch/pull/85382 on this model?

**Comment by anijain2305** [comment]
I tried the latest master (where the above PR is merged), but the accuracy issue still exists

**Comment by ezyang** [comment]
Does the minifier report accuracy failure even when it retains the whole graph? Can it dump that out as a "minimal" repro?

**Comment by anijain2305** [comment]
I didn't fully understand the question.

In this models, there are many graph breaks. So there are multiple of tens of subgraphs.

For each subgraph, minifier compares eager vs dynamo accuracy. If it fails, it starts minifying that subgraph further.

The issue here is that accuracy minifier does not find any subgraph offending. (even though the final accuracy fails)

**Comment by ezyang** [comment]
So... tighten the bounds?

**Comment by voznesenskym** [comment]
> In this model, there are many graph breaks. So there are multiple of tens of subgraphs.
> 
> For each subgraph, minifier compares eager vs dynamo accuracy. If it fails, it dumps the offending subgraph and starts minifying it further (dumping at every success of minification).
> 
> The issue here is that accuracy minifier does not find any subgraph offending. (even though the final accuracy fails)

noob question: How is that possible? Presumably we are evaluating eager correctly outside of graphs, so either:

1) The outputs of graphs are uniform and not offending, so the inputs to eager are identical, so the sum total of graphs + eager is non offending
2) The outputs of graphs are non uniform and offending, so the inputs to eager are not identical, and then there is ostensibly some graph that has bad outputs (aka bad inputs to the next eager phase), and that is the offending graph

But how can it be that we have "The outputs of graphs are uniform and not offending" from 1, but then fail accuracy after?

**Comment by eellison** [comment]
I dumped the graphs with / without fake tensors here: https://www.diffchecker.com/39SIy56d

**Comment by anijain2305** [comment]
I think the issue is how we check accuracy in minifier. 

For each subgraph, we check the forward output and gradients. For backward, we just do `sum` of the output tensor, and do `.backward()` call.

In actual run, the inputs to the backward graphs can have different strides. And this deviates the accuracy checking in minifier vs real run.

This inconsistency is not-so-easy to resolve. During the bytecode analysis, we have fx Graph and just the forward `example_inputs`. Let me think if there is a way to delay this and run minifier with actual inputs to the backward pass.

**Comment by eellison** [comment]
Note that without functionalization - the inaccuracy is pretty minimal - `torchdynamo.utils: [ERROR] RMSE (res-fp64): 0.01575, (ref-fp64): 0.00002`

**Comment by anijain2305** [comment]
> Note the above graph is without functionalization, and the inaccuracy is pretty minimal - `torchdynamo.utils: [ERROR] RMSE (res-fp64): 0.01575, (ref-fp64): 0.00002`

This inaccuracy seems pretty high to me. I would `ref` and `ref` to be very different here.

**Comment by eellison** [comment]
Yea, you're right, I should say, smaller than it is without functionalization.

**Comment by ezyang** [comment]
We're running eager though, not inductor, so if it's possible to run the model eagerly against itself, and be bitwise deterministic, then it should be possible to run aot eager with bitwise determinism too

**Comment by anijain2305** [comment]
I debugged little more. I was able to narrow it down to 2 subgraphs. This [patch](https://www.internalfb.com/phabricator/paste/view/P532175289) sends those 2 subgraphs to AotAutograd, and rest fallback to eager.

As we already know, disabling fake tensors remove this error. So, I also diffed the joint_graph w and w/o fake tensor. Here is the [diff](https://www.diffchecker.com/4hza2INc). Funny thing - fake tensors don't have an entire mm op.
